File: apps/handover/pivot/Pivot.Client.Plugin/src/client-manifest.ts
import type { ContentLocation } from './content.js';


/**
 * Route declaration from a plugin's client manifest.
 * Describes a route the plugin contributes to the app shell.
 */
export interface PluginRouteDeclaration {
	/** URL path pattern (e.g. `/weather`, `/weather/:city`). */
	path:          string;
	/** Unique route name for programmatic navigation. */
	name:          string;
	/** Display label for navigation menus. */
	label?:        string;
	/** Icon identifier for navigation menus. */
	icon?:         string;
	/**
	 * Export name of the component in the plugin's entry module.
	 * Resolved via dynamic import at navigation time.
	 */
	lazyComponent: string;
}


/**
 * Content area declaration from a plugin's client manifest.
 * Describes a UI panel/sidebar/editor area the plugin contributes.
 */
export interface PluginContentDeclaration {
	/** Unique content area identifier. */
	id:                 string;
	/** Where the content appears by default. */
	defaultLocation:    ContentLocation;
	/** All locations where this content can be placed. */
	availableLocations: ContentLocation[];
	/** Tab display metadata. */
	tab: {
		id:    string;
		title: string;
		icon:  string;
	};
	/**
	 * Export name of the content component in the plugin's entry module.
	 * Resolved via dynamic import when the content area is activated.
	 */
	lazyComponent: string;
}


/**
 * Service declaration from a plugin's client manifest.
 * Describes a service the plugin registers into the DI container on activation.
 */
export interface PluginServiceDeclaration {
	/** DI identifier to register the service under. */
	identifier: string;
	/** Export name of the service factory/class in the plugin's entry module. */
	exportName: string;
}


/**
 * Statusbar contribution from a plugin's client manifest.
 */
export interface PluginStatusbarDeclaration {
	/** Unique identifier. */
	id:            string;
	/** Alignment in the statusbar. */
	alignment:     'left' | 'right';
	/** Priority (higher = closer to edge). */
	priority?:     number;
	/** Export name of the statusbar component in the plugin's entry module. */
	lazyComponent: string;
}


/**
 * Shared dependency declaration.
 * Specifies a package the plugin expects the host to provide.
 */
export interface SharedDependency {
	/** npm package name (e.g. `lit`, `@arcmantle/injector`). */
	name:         string;
	/** Semver range the plugin is compatible with (e.g. `^3.0.0`). */
	versionRange: string;
}


/**
 * The client-side manifest for a Pivot plugin.
 * Serialized as `client-manifest.json` in the plugin's `/client/` directory.
 *
 * This manifest is loaded eagerly by the app shell on boot.
 * Actual plugin code is loaded lazily when routes or content areas are activated.
 */
export interface ClientManifest {
	/** Relative path to the ES module entry point (e.g. `index.js`). */
	entryModule:   string;
	/** Optional CSS files to load with the plugin. */
	styles?:       string[];
	/** Routes the plugin contributes to the app shell. */
	routes?:       PluginRouteDeclaration[];
	/** Content areas the plugin contributes to the layout. */
	contentAreas?: PluginContentDeclaration[];
	/** Services the plugin registers on activation. */
	services?:     PluginServiceDeclaration[];
	/** Statusbar items the plugin contributes. */
	statusbar?:    PluginStatusbarDeclaration[];
	/** Dependency declarations. */
	dependencies?: {
		/** Packages the plugin expects the host to provide via import map. */
		shared?:  Record<string, string>;
		/** Packages the plugin bundles itself (informational). */
		bundled?: string[];
	};
	/**
	 * Pre-built shared dependency bundles shipped by this plugin.
	 *
	 * Each entry maps a root package name to its exact version and the
	 * list of specifier-to-file mappings within the `shared/` output directory.
	 *
	 * During startup the backend resolves version conflicts across all
	 * plugins (highest version per major wins) and serves the winning
	 * plugin's bundles at `/shared/`.
	 *
	 * Generated automatically by the Vite plugin at build time.
	 */
	sharedBundles?: Record<string, SharedBundle>;
}


/**
 * Describes a pre-built shared dependency bundle shipped with a plugin.
 */
export interface SharedBundle {
	/** Exact installed version of the root package (e.g. `3.2.1`). */
	version: string;
	/**
	 * Maps each bare specifier to its output file path relative to
	 * the plugin's `/client/shared/` directory.
	 *
	 * @example
	 * ```json
	 * {
	 *   "lit":                   "lit.js",
	 *   "lit/decorators.js":     "lit__decorators.js",
	 *   "lit/directives/when.js": "lit__directives__when.js"
	 * }
	 * ```
	 */
	files:   Record<string, string>;
}


/**
 * A resolved plugin descriptor returned by the backend API.
 * Combines the plugin identity with its client manifest and asset base URL.
 */
export interface ResolvedPluginDescriptor {
	/** Plugin name (matches the backend PluginManifest.Name). */
	name:           string;
	/** Semantic version of the plugin. */
	version:        string;
	/** Base URL where the plugin's client assets are served. */
	baseUrl:        string;
	/** The plugin's client manifest. */
	clientManifest: ClientManifest;
}

File: pivot/Pivot.Core/Plugin/ClientManifest.cs
using System.Text.Json.Serialization;

namespace Pivot.Plugin;


/// <summary>
/// Client-side manifest for a Pivot plugin.
/// Deserialized from `client-manifest.json` in the plugin's `/client/` directory.
/// Describes what the plugin contributes to the app shell (routes, content areas, services).
/// </summary>
public class ClientManifest {
	/// <summary>
	/// Relative path to the ES module entry point (e.g. "index.js").
	/// </summary>
	[JsonPropertyName("entryModule")]
	public required string EntryModule { get; set; }

	/// <summary>
	/// Optional CSS files to load with the plugin.
	/// </summary>
	[JsonPropertyName("styles")]
	public List<string>? Styles { get; set; }

	/// <summary>
	/// Routes the plugin contributes to the app shell.
	/// </summary>
	[JsonPropertyName("routes")]
	public List<ClientPluginRoute>? Routes { get; set; }

	/// <summary>
	/// Content areas the plugin contributes to the app shell layout.
	/// </summary>
	[JsonPropertyName("contentAreas")]
	public List<ClientPluginContentArea>? ContentAreas { get; set; }

	/// <summary>
	/// Services the plugin registers into the DI container on activation.
	/// </summary>
	[JsonPropertyName("services")]
	public List<ClientPluginService>? Services { get; set; }

	/// <summary>
	/// Statusbar items the plugin contributes.
	/// </summary>
	[JsonPropertyName("statusbar")]
	public List<ClientPluginStatusbarItem>? Statusbar { get; set; }

	/// <summary>
	/// Dependency declarations.
	/// </summary>
	[JsonPropertyName("dependencies")]
	public ClientPluginDependencies? Dependencies { get; set; }

	/// <summary>
	/// Pre-built shared dependency bundles shipped by this plugin.
	/// Key = root package name (e.g. "lit"), Value = bundle metadata.
	///
	/// During startup the backend resolves version conflicts across all
	/// plugins (highest version per major wins) and serves the winning
	/// plugin's bundles at /shared/.
	///
	/// Generated automatically by the Vite plugin at build time.
	/// </summary>
	[JsonPropertyName("sharedBundles")]
	public Dictionary<string, SharedBundle>? SharedBundles { get; set; }
}


/// <summary>
/// Describes a pre-built shared dependency bundle shipped with a plugin.
/// </summary>
public class SharedBundle {
	/// <summary>
	/// Exact installed version of the root package (e.g. "3.2.1").
	/// </summary>
	[JsonPropertyName("version")]
	public required string Version { get; set; }

	/// <summary>
	/// Maps each bare specifier to its output file path relative to
	/// the plugin's /client/shared/ directory.
	/// E.g. { "lit": "lit.js", "lit/decorators.js": "lit__decorators_js.js" }
	/// </summary>
	[JsonPropertyName("files")]
	public required Dictionary<string, string> Files { get; set; }
}


/// <summary>
/// A route declaration from a plugin's client manifest.
/// </summary>
public class ClientPluginRoute {
	[JsonPropertyName("path")]
	public required string Path { get; set; }

	[JsonPropertyName("name")]
	public required string Name { get; set; }

	[JsonPropertyName("label")]
	public string? Label { get; set; }

	[JsonPropertyName("icon")]
	public string? Icon { get; set; }

	[JsonPropertyName("lazyComponent")]
	public required string LazyComponent { get; set; }
}


/// <summary>
/// A content area declaration from a plugin's client manifest.
/// </summary>
public class ClientPluginContentArea {
	[JsonPropertyName("id")]
	public required string Id { get; set; }

	[JsonPropertyName("defaultLocation")]
	public required string DefaultLocation { get; set; }

	[JsonPropertyName("availableLocations")]
	public required List<string> AvailableLocations { get; set; }

	[JsonPropertyName("tab")]
	public required ClientPluginTab Tab { get; set; }

	[JsonPropertyName("lazyComponent")]
	public required string LazyComponent { get; set; }
}


/// <summary>
/// Tab metadata for a content area.
/// </summary>
public class ClientPluginTab {
	[JsonPropertyName("id")]
	public required string Id { get; set; }

	[JsonPropertyName("title")]
	public required string Title { get; set; }

	[JsonPropertyName("icon")]
	public required string Icon { get; set; }
}


/// <summary>
/// A service declaration from a plugin's client manifest.
/// </summary>
public class ClientPluginService {
	[JsonPropertyName("identifier")]
	public required string Identifier { get; set; }

	[JsonPropertyName("exportName")]
	public required string ExportName { get; set; }
}


/// <summary>
/// A statusbar item declaration from a plugin's client manifest.
/// </summary>
public class ClientPluginStatusbarItem {
	[JsonPropertyName("id")]
	public required string Id { get; set; }

	[JsonPropertyName("alignment")]
	public required string Alignment { get; set; }

	[JsonPropertyName("priority")]
	public int? Priority { get; set; }

	[JsonPropertyName("lazyComponent")]
	public required string LazyComponent { get; set; }
}


/// <summary>
/// Dependency declarations for a client plugin.
/// </summary>
public class ClientPluginDependencies {
	/// <summary>
	/// Packages the plugin expects the host to provide via import map.
	/// Key = package name, Value = semver range.
	/// </summary>
	[JsonPropertyName("shared")]
	public Dictionary<string, string>? Shared { get; set; }

	/// <summary>
	/// Packages the plugin bundles itself (informational only).
	/// </summary>
	[JsonPropertyName("bundled")]
	public List<string>? Bundled { get; set; }
}


/// <summary>
/// Resolved plugin descriptor returned by the client-manifests API.
/// Combines plugin identity with its client manifest and asset base URL.
/// </summary>
public class ResolvedClientPlugin {
	[JsonPropertyName("name")]
	public required string Name { get; set; }

	[JsonPropertyName("version")]
	public required string Version { get; set; }

	[JsonPropertyName("baseUrl")]
	public required string BaseUrl { get; set; }

	[JsonPropertyName("clientManifest")]
	public required ClientManifest ClientManifest { get; set; }
}
